File: java/src/main/java/pl/adambanaszkiewicz/four_layers_example/infrastructure_layer/to_this/DownloadManager.java
package pl.adambanaszkiewicz.four_layers_example.infrastructure_layer.to_this;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import pl.adambanaszkiewicz.four_layers_example.infrastructure_layer.to_this.FilesFinderInterface;
import pl.adambanaszkiewicz.four_layers_example.infrastructure_layer.to_this.FilesystemInterface;

import java.util.Map;

@Component
public class DownloadManager {

    private final FilesFinderInterface filesFinder;
    private final FilesystemInterface filesystem;

    @Autowired
    public DownloadManager(
        FilesFinderInterface filesFinder,
        FilesystemInterface filesystem
    ) {
        this.filesFinder = filesFinder;
        this.filesystem = filesystem;
    }

    public boolean fileExistsOnServer(int fileId) {
        FileDTO file = filesFinder.find(fileId);

        if (file == null) {
            return false;
        }

        // Business logic...

        return filesystem.exists(file.filepath);
    }
}

File: php/src/InfrastructureLayer/ToThis/DownloadManager.php
<?php

declare(strict_types=1);

namespace App\InfrastructureLayer\ToThis;

final class DownloadManager
{
    public function __construct(
        private readonly FilesFinderInterface $filesFinder,
        private readonly FilesystemInterface $filesystem,
    ) {
    }

    public function fileExistsOnServer(int $fileId): bool
    {
        $file = $this->filesFinder->find($fileId);

        if (!$file) {
            return false;
        }

        /** Business logic... */

        return $this->filesystem->exists($file['filepath']);
    }
}

